hashing/src/indexed_merkle_proof.rs

```diff
@@ -0,0 +1,525 @@
+            if leaf_count == 0 {
+                return Digest::SENTINEL_MERKLE_TREE;
+            }
+            if leaf_count == 1 {
+                return proof[0];
+            }
+            let half = 1u64 << (63 - (leaf_count - 1).leading_zeros());
+            let last = proof.len() - 1;
+            if index < half {
+                let left = compute_raw_root_from_proof(index, half, &proof[..last]);
+                Digest::hash_pair(&left, &proof[last])
+            } else {
+                let right =
+                    compute_raw_root_from_proof(index - half, leaf_count - half, &proof[..last]);
+                Digest::hash_pair(&proof[last], &right)
+            }
+        }
+
+        let raw_root = compute_raw_root_from_proof(index, count, proof);
+        Digest::hash_pair(count.to_le_bytes(), raw_root)
+    }
+
+    /// Construct an `IndexedMerkleProof` with a proof of zero digests.
+    fn test_indexed_merkle_proof(index: u64, count: u64) -> IndexedMerkleProof {
+        let mut indexed_merkle_proof = IndexedMerkleProof {
+            index,
+            count,
+            merkle_proof: vec![],
+        };
+        let expected_proof_length = indexed_merkle_proof.compute_expected_proof_length();
+        indexed_merkle_proof.merkle_proof = std::iter::repeat_with(|| Digest([0u8; 32]))
+            .take(expected_proof_length as usize)
+            .collect();
+        indexed_merkle_proof
+    }
+
+    #[proptest]
+    fn root_from_proof_agrees_with_recursion(index: u64, count: u64) {
+        let indexed_merkle_proof = test_indexed_merkle_proof(index, count);
+        prop_assert_eq!(
+            indexed_merkle_proof.root_hash(),
+            reference_root_from_proof(
+                indexed_merkle_proof.index,
+                indexed_merkle_proof.count,
+                indexed_merkle_proof.merkle_proof(),
+            ),
+            "Result did not agree with reference implementation.",
+        );
+    }
+
+    #[test]
+    fn root_from_proof_agrees_with_recursion_2147483648_4294967297() {
+        let indexed_merkle_proof = test_indexed_merkle_proof(2147483648, 4294967297);
+        assert_eq!(
+            indexed_merkle_proof.root_hash(),
+            reference_root_from_proof(
+                indexed_merkle_proof.index,
+                indexed_merkle_proof.count,
+                indexed_merkle_proof.merkle_proof(),
+            ),
+            "Result did not agree with reference implementation.",
+        );
+    }
+
+    #[test]
+    fn serde_deserialization_of_malformed_proof_should_work() {
+        let indexed_merkle_proof = test_indexed_merkle_proof(10, 10);
+
+        let json = serde_json::to_string(&indexed_merkle_proof).unwrap();
+        assert_eq!(
+            indexed_merkle_proof,
+            serde_json::from_str::<IndexedMerkleProof>(&json)
+                .expect("should deserialize correctly")
+        );
+
+        // Check that proof with index greater than count fails to deserialize
```

Comment:
This and the next three comments are outdated, are they?

Reply:
Yes, fixed.